packages/next/src/cli/next-dev.ts

```diff
@@ -98,9 +98,71 @@ if (!isChildProcess) {
   process.on('SIGTERM', () => process.exit(0))
 }
 
-function watchConfigFiles(dirToWatch: string) {
+/** Preven't race conditions on async function, by waiting for the previous invocation to finish */
+const functionNotRunningInParallel = (fn: () => Promise<void>) => {
+  const state = {
+    running: false,
+    runWhenDone: false,
+  }
+  const wrappedFn = async () => {
+    if (state.running) {
+      state.runWhenDone = true
+      return
+    }
+    state.running = true
+    try {
+      await fn()
+    } finally {
+      state.running = false
+    }
+
+    if (state.runWhenDone) {
+      state.runWhenDone = false
+      wrappedFn()
+    }
+  }
+  return wrappedFn
+}
+
+/**
+ * Run effect on config change, and return a function to stop watching.
+ */
+const restartedOnConfigChange = async (
+  dirToWatch: string,
+  fn: () => Promise<() => Promise<void>>
+) => {
+  if (unwatchConfigFiles) {
+    await unwatchConfigFiles()
+  }
+
+  let stop = async () => {}
+  const restart = functionNotRunningInParallel(async () => {
+    await stop()
+    stop = await fn()
+  })
+
+  const wp = new Watchpack()
+  wp.watch({ files: CONFIG_FILES.map((file) => path.join(dirToWatch, file)) })
+  wp.on('change', async (filename) => {
```

Comment:
We should be checking for actual hash changes as well similar to `instrumentation` as change events can fire without underlying content change and a dev server reboot isn't cheap.

Reply:
if you're talking about rebooting, yes. Instrumentation libraries are usually monkey-patching a lot of things so it's better to reload early and start from a fresh instance. Otherwise that's a memory leak fest

Reply:
Oh, sorry I was talking about checking the hash.
It's not that much extra work, I was just surprised that it made any difference.